### PR TITLE
Employee apis integration

### PR DESCRIPTION
**Pull request by jkhabra**
Resolve #

### CODE REVIEW

File: ui/src/components/Employees/List.tsx [hunk 1]
```diff
@@ -1,41 +1,34 @@
...
-  id: string;
-  name: string;
-  headShot?: string;
-  organization: any[];
-  designation: string;
-}
-
 interface FieldProps {
-  list: EmployeeList[];
+  list: Employee[];
```
Comment (bitspook) [inline ui/src/components/Employees/List.tsx]: Please change `list` to `employees`

File: ui/src/components/Employees/List.tsx [hunk 2]
```diff
@@ -1,41 +1,34 @@
...
-    </span>
-  );
-};
+/* const organization = (org: any) => {
+ *   return (
+ *     <span className={c('project-items')} key={org}>
+ *       {org}
+ *     </span>
+ *   );
+ * }; */
```
Comment (bitspook) [inline ui/src/components/Employees/List.tsx]: Please don't leave commented out code.

File: ui/src/components/Organization/List.tsx [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 import style from 'src/components/Organization/style.module.scss';
 
-import { Organization } from 'src/entities/Organizations';
+import { Employee } from 'src/entities/Organizations';
 
 const c = classNames.bind(style);
 
 interface OrganizationProps {
-  list: Organization[];
+  list: Employee[];
```
Comment (bitspook) [inline ui/src/components/Organization/List.tsx]: Type should be `Organization[]`, prop should be named `organizations`.

File: ui/src/duck/designation.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+
+import { Designation, fetch } from 'src/entities/Designation';
```
Comment (bitspook) [inline ui/src/duck/designation.ts]: Let's rename to `fetch` to `fetchAll`. `fetch` being a global in browser might trip someone

File: ui/src/duck/designation.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  error?: string;
+  isLoading: boolean;
+  designation: Designation[];
+}
+
+export const fetchDesignation = createAsyncThunk<
+  Designation[],
+undefined,
+{ rejectValue: Error; state: DesignationState }
+  >('designations/fetch', fetch);
```
Comment (bitspook) [inline ui/src/duck/designation.ts]: Let's start following a convention of naming fetch action as `fetchAll` and `fetchOne`.

File: ui/src/duck/designation.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+
+import { Designation, fetch } from 'src/entities/Designation';
+
+export interface DesignationState {
+  error?: string;
```
Comment (bitspook) [inline ui/src/duck/designation.ts]: Let's get rid of error if we are not using it. It'll only end up misleading someone that we are handling this error.

File: ui/src/duck/employee.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+
+import { create, CreatePayload, Employee, fetch } from 'src/entities/Employee';
+
+export interface EmployeeState {
+  employees: Employee[];
+  error?: string;
```
Comment (bitspook) [inline ui/src/duck/employee.ts:7]: Until we start using this error, let's get rid of this.

File: ui/src/entities/Designation.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  name: string;
+}
+
+export const fetch = async (): Promise<Designation[]> => {
+  const query = `{
+                 designations {
+                 id
+                 name
+                 }
+               }`;
```
Comment (bitspook) [inline ui/src/entities/Designation.ts]: Please fix the indentation

File: ui/src/entities/Employee.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+               })
+               {
+                ecode
+                name
+                email
+                headshot
+                joining_date
+                designation_id
+            }
+          }`;
```
Comment (bitspook) [inline ui/src/entities/Employee.ts]: Please fix indentation.

File: ui/src/entities/Employee.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+import client from 'src/lib/client';
+
+export interface Employee {
+  ecode: string;
+  email: string;
+  name: string;
+  headshot: string;
+  joining_date: string;
+  designation_id: string;
```
Comment (bitspook) [inline ui/src/entities/Employee.ts:11]: Please also add organiztion_id in employee

File: ui/src/pages/employees/add.tsx [hunk 11]
```diff
@@ -1,41 +1,64 @@
...
-  { id: '2', name: 'De', avtar: '' },
-  { id: '3', name: 'Bd', avtar: '' },
-];
+const AddEmployee: React.FC<FormikProps<FormValues>> = (p) => {
+  const dispatch = useDispatch();
+  const { organizations, isFetching } = useSelector<RootState, OrganizationState>(
+    (state) => state.organization,
+  );
+
+  const { designation, isLoading } = useSelector<RootState, DesignationState>(
```
Comment (bitspook) [inline ui/src/pages/employees/add.tsx]: Let's rename similarly named variables and give more context specific names. e.g `isFetching` can become `isFetchingOrganizations`, and `isLoading` can become `isFetchingDesignations`.

File: ui/src/pages/employees/add.tsx [hunk 12]
```diff
@@ -83,39 +106,64 @@ const AddOrg: React.FC<FormikProps<EmployeeFormValues>> = (props) => {
...
+    const resp = await dispatch(createEmployee(values));
+
+    actions.setSubmitting(false);
+
+    try {
+      unwrapResult(resp as any);
+    } catch (err) {
+      /* if (/uniqueness violation/i.test(err.message)) {
+       *   actions.setFieldError('email', 'An organization with same name already exists');
+       * } */
```
Comment (bitspook) [inline ui/src/pages/employees/add.tsx]: Please don't leave commented out code.

**Review by bitspook — CHANGES_REQUESTED** [review]

File: ui/src/components/Form/SelectField/style.module.scss [hunk 13]
```diff
@@ -33,6 +33,8 @@
 }
 
 .select-options {
+  height: 12em;
+  overflow-y: scroll;
```
Comment (bitspook) [inline ui/src/components/Form/SelectField/style.module.scss]: Just fyi, you can also use `overflow-y: auto`.

File: ui/src/duck/designation.ts [hunk 14]
```diff
@@ -1,37 +1,31 @@
...
+  isFetching: boolean;
   designation: Designation[];
 }
 
 export const fetchDesignation = createAsyncThunk<
   Designation[],
 undefined,
 { rejectValue: Error; state: DesignationState }
-  >('designations/fetch', fetch);
+  >('designations/fetch/many', fetchMany);
```
Comment (bitspook) [inline ui/src/duck/designation.ts]: Using `/` is akin to adding a namespace. `fetch/many` doesn't look like a valid case for using a namespace. I recommend we make it `designations/fetchMany` (and `designations/fetchOne`)

**Review by bitspook — CHANGES_REQUESTED** [review]
Looks good overall. Just one minor comment about namespacing action names.

File: ui/src/components/Form/SelectField/style.module.scss (hunk 13 above)
Comment (jkhabra) [reply ui/src/components/Form/SelectField/style.module.scss]: ok

File: ui/src/duck/designation.ts (hunk 14 above)
Comment (jkhabra) [reply ui/src/duck/designation.ts]: alright that's make sense

**Review by bitspook — APPROVED** [review]

**Pull request merged**